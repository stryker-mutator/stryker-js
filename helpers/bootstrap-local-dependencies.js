// @ts-check
import { LocalInstaller, progress } from 'install-local';
import fs from 'fs';
import glob from 'glob';
import path from 'path';

console.log('starting installation of local dependencies');

/**
 * @param {string} pattern 
 * @param {glob.IOptions} options 
 * @returns {Promise<string[]>}
 */
function globAsPromised(pattern, options) {
  return new Promise((res, rej) => {
    glob(pattern, options, (err, matches) => {
      if (err) {
        rej(err);
      } else {
        res(matches);
      }
    });
  });
}

/**
 * @typedef Package
 * @property {{ [name: string]: string }} localDependencies

/**
 * Installs local dependencies in one go,
 * reads the package.json files and installs the local dependencies marked there.
 * The packageJsonGlobs default to ./package.json and test/* /package.json files 
 * @param {string} directory
 * @param {string[]} packageJsonGlobs
 * @returns {Promise<void>}
 */
export async function bootstrapLocalDependencies(directory, packageJsonGlobs = ['package.json', 'test/*/package.json']) {
  console.log(`bootstrap ${path.resolve(directory)} (using ${packageJsonGlobs.join(',')})`);
  const files = (await Promise.all(packageJsonGlobs.map((globPattern) => globAsPromised(globPattern, { cwd: path.resolve(directory) })))).flat();
  const packages = await Promise.all(files.map(fileName => fs.promises.readFile(fileName, 'utf-8')
    .then((rawContent) => {
      /**
       * @type {Package}
       */
      const content = JSON.parse(rawContent);
      return { dir: path.dirname(fileName), content };
    })));
  /**
   * @type {import('install-local').ListByPackage}
   */
  const sourcesByTarget = {};
  for (const pkg of packages) {
    const localDeps = pkg.content.localDependencies;
    if (localDeps) {
      sourcesByTarget[path.resolve(pkg.dir)] = Object.keys(localDeps).map(key => path.resolve(pkg.dir, localDeps[key]));
    }
  }
  const localInstaller = new LocalInstaller(sourcesByTarget);
  progress(localInstaller);
  await localInstaller.install();
}
