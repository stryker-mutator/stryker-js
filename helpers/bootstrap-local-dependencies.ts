import { ListByPackage, LocalInstaller, progress } from 'install-local';
import * as fs from 'fs';
import glob = require('glob');
import path = require('path');

console.log('starting installation of local dependencies');

function globAsPromised(pattern: string, options: glob.IOptions) {
  return new Promise<string[]>((res, rej) => {
    glob(pattern, options, (err, matches) => {
      if (err) {
        rej(err);
      } else {
        res(matches);
      }
    });
  });
}

function readFile(fileName: string) {
  return new Promise<string>((res, rej) => fs.readFile(fileName, 'utf8', (err, content) => {
    if (err) {
      rej(err);
    } else {
      res(content);
    }
  }));
}

interface Package {
  localDependencies?: { [name: string]: string };
}

/**
 * Installs local dependencies in one go,
 * reads package.json and test/* /package.json files and installs the local dependencies marked there
 * @param directory the directory where the tests live
 */
export async function bootstrapLocalDependencies(directory: string) {
  console.log('bootstrap ' + path.resolve(directory));
  const files = await globAsPromised('{package.json,test/*/package.json}', { cwd: path.resolve(directory) });
  const packages = await Promise.all(files.map(fileName => readFile(fileName)
    .then(content => ({ dir: path.dirname(fileName), content: JSON.parse(content) as Package }))));
  const sourcesByTarget: ListByPackage = {};
  for (const pkg of packages) {
    const localDeps = pkg.content.localDependencies;
    if (localDeps) {
      sourcesByTarget[path.resolve(pkg.dir)] = Object.keys(localDeps).map(key => path.resolve(pkg.dir, localDeps[key]));
    }
  }
  const localInstaller = new LocalInstaller(sourcesByTarget);
  progress(localInstaller);
  await localInstaller.install();
}
