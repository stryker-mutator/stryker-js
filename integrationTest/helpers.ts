import * as path from 'path';
import * as fs from 'mz/fs';
import { expect } from 'chai';
import { ScoreResult } from 'stryker-api/report';

export async function readScoreResult(eventResultDirectory = path.resolve('reports', 'mutation', 'events')) {
    const allReportFiles = await fs.readdir(eventResultDirectory);
    const scoreResultReportFile = allReportFiles.find(file => !!file.match(/.*onScoreCalculated.*/));
    expect(scoreResultReportFile).ok;
    const scoreResultContent = await fs.readFile(path.resolve(eventResultDirectory, scoreResultReportFile || ''), 'utf8');
    return JSON.parse(scoreResultContent) as ScoreResult;
}