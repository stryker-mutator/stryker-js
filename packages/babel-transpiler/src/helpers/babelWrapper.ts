/**
 * This wrapper is needed because babel/core only exports
 * non-stubbable es6 properties. See node_modules/@babel/core/lib/index.js
 */
export * from '@babel/core';
