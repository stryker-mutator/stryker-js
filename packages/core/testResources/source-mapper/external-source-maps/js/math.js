"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function max() {
    var numbers = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        numbers[_i] = arguments[_i];
    }
    return Math.max.apply(Math, numbers);
}
exports.max = max;
function total() {
    var numbers = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        numbers[_i] = arguments[_i];
    }
    return numbers.reduce(function (a, b) { return a + b; });
}
exports.total = total;
//# sourceMappingURL=math.js.map