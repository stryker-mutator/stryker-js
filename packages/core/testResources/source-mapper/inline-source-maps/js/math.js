"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function max() {
    var numbers = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        numbers[_i] = arguments[_i];
    }
    return Math.max.apply(Math, numbers);
}
exports.max = max;
function total() {
    var numbers = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        numbers[_i] = arguments[_i];
    }
    return numbers.reduce(function (a, b) { return a + b; });
}
exports.total = total;
//# sourceMappingURL=data:application/json;base64,eyJ2ZXJzaW9uIjozLCJmaWxlIjoibWF0aC5qcyIsInNvdXJjZVJvb3QiOiIiLCJzb3VyY2VzIjpbIi4uL3RzL3NyYy9tYXRoLnRzIl0sIm5hbWVzIjpbXSwibWFwcGluZ3MiOiI7O0FBRUE7SUFBb0IsaUJBQW9CO1NBQXBCLFVBQW9CLEVBQXBCLHFCQUFvQixFQUFwQixJQUFvQjtRQUFwQiw0QkFBb0I7O0lBQ3RDLE1BQU0sQ0FBQyxJQUFJLENBQUMsR0FBRyxPQUFSLElBQUksRUFBUSxPQUFPLEVBQUU7QUFDOUIsQ0FBQztBQUZELGtCQUVDO0FBRUQ7SUFBc0IsaUJBQW9CO1NBQXBCLFVBQW9CLEVBQXBCLHFCQUFvQixFQUFwQixJQUFvQjtRQUFwQiw0QkFBb0I7O0lBQ3hDLE1BQU0sQ0FBQyxPQUFPLENBQUMsTUFBTSxDQUFDLFVBQUMsQ0FBQyxFQUFFLENBQUMsSUFBSyxPQUFBLENBQUMsR0FBRyxDQUFDLEVBQUwsQ0FBSyxDQUFDLENBQUM7QUFDekMsQ0FBQztBQUZELHNCQUVDIn0=