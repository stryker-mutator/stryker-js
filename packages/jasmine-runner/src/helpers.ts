import { TestResult, TestStatus } from '@stryker-mutator/api/test-runner';
import jasmine from 'jasmine';

export const helpers = {
  toStrykerTestResult(specResult: jasmine.SpecResult, timeSpentMs: number): TestResult {
    const baseResult = {
      id: specResult.id.toString(),
      name: specResult.fullName,
      timeSpentMs,
    };
    if (specResult.status === 'disabled' || specResult.status === 'pending' || specResult.status === 'excluded') {
      return {
        ...baseResult,
        status: TestStatus.Skipped,
      };
    } else if (!specResult.failedExpectations || specResult.failedExpectations.length === 0) {
      return {
        ...baseResult,
        status: TestStatus.Success,
      };
    } else {
      return {
        ...baseResult,
        status: TestStatus.Failed,
        failureMessage: specResult.failedExpectations.map((failedExpectation) => failedExpectation.message).join(','),
      };
    }
  },
  createJasmine(options: jasmine.JasmineOptions): jasmine {
    return new jasmine(options);
  },
};
