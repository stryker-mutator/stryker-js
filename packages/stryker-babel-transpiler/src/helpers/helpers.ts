import * as path from 'path';

export function toJSFileName(fileName: string) {
  const fileNameInfo = path.parse(fileName);
  const newFileName = path.join(fileNameInfo.dir, fileNameInfo.name + '.js');
  return newFileName;
}
