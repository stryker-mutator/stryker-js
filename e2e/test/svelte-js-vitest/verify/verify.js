// @ts-check
import { expect } from 'chai';

import { expectMetricsJsonToMatchSnapshot, readMutationTestingJsonResult } from '../../../helpers.js';

/** @param {number} n */
function pad(n) {
  return n.toString().padStart(2, '0');
}

/**
 * @param {import('mutation-testing-report-schema').MutantResult} mutant
 * @returns {string}
 */
function mutantKey(mutant) {
  return `${pad(mutant.location.start.line)}:${pad(mutant.location.start.column)}-${pad(mutant.location.end.line)}:${pad(
    mutant.location.end.column,
  )}-${mutant.mutatorName}-${mutant.replacement}`;
}

describe('Verify stryker has ran correctly', () => {
  it('should report correct score', async () => {
    await expectMetricsJsonToMatchSnapshot();
  });

  it('should report mutants at the correct locations', async () => {
    const actual = await readMutationTestingJsonResult();
    const app = actual.files['src/App.svelte'];
    const locatedMutants = [...app.mutants]
      .sort((a, b) => mutantKey(a).localeCompare(mutantKey(b)))
      .map(({ mutatorName, location: { start, end }, replacement }) => ({
        mutatorName,
        location: `${start.line}:${start.column}-${end.line}:${end.column}`,
        replacement,
      }));
    expect(locatedMutants).matchSnapshot();
  });
});
