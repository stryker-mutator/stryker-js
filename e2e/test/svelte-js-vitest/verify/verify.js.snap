// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Verify stryker has ran correctly should report correct score 1`] = `
Object {
  "compileErrors": 0,
  "ignored": 0,
  "killed": 13,
  "mutationScore": 64,
  "mutationScoreBasedOnCoveredCode": 84.21052631578947,
  "noCoverage": 6,
  "pending": 0,
  "runtimeErrors": 0,
  "survived": 3,
  "timeout": 3,
  "totalCovered": 19,
  "totalDetected": 16,
  "totalInvalid": 0,
  "totalMutants": 25,
  "totalUndetected": 9,
  "totalValid": 25,
}
`;

exports[`Verify stryker has ran correctly should report mutants at the correct locations 1`] = `
Array [
  Object {
    "location": "2:29-4:4",
    "mutatorName": "BlockStatement",
    "replacement": "{}",
  },
  Object {
    "location": "3:12-3:17",
    "mutatorName": "ArithmeticOperator",
    "replacement": "a - b",
  },
  Object {
    "location": "6:34-8:4",
    "mutatorName": "BlockStatement",
    "replacement": "{}",
  },
  Object {
    "location": "7:12-7:17",
    "mutatorName": "ArithmeticOperator",
    "replacement": "a / b",
  },
  Object {
    "location": "9:36-11:4",
    "mutatorName": "BlockStatement",
    "replacement": "{}",
  },
  Object {
    "location": "10:12-10:21",
    "mutatorName": "ConditionalExpression",
    "replacement": "false",
  },
  Object {
    "location": "10:12-10:21",
    "mutatorName": "ConditionalExpression",
    "replacement": "true",
  },
  Object {
    "location": "10:12-10:21",
    "mutatorName": "EqualityOperator",
    "replacement": "age < 18",
  },
  Object {
    "location": "10:12-10:21",
    "mutatorName": "EqualityOperator",
    "replacement": "age > 18",
  },
  Object {
    "location": "15:35-22:4",
    "mutatorName": "BlockStatement",
    "replacement": "{}",
  },
  Object {
    "location": "16:16-16:20",
    "mutatorName": "BooleanLiteral",
    "replacement": "false",
  },
  Object {
    "location": "17:12-17:16",
    "mutatorName": "ConditionalExpression",
    "replacement": "false",
  },
  Object {
    "location": "17:18-21:6",
    "mutatorName": "BlockStatement",
    "replacement": "{}",
  },
  Object {
    "location": "19:7-19:10",
    "mutatorName": "UpdateOperator",
    "replacement": "n++",
  },
  Object {
    "location": "20:14-20:19",
    "mutatorName": "ConditionalExpression",
    "replacement": "false",
  },
  Object {
    "location": "20:14-20:19",
    "mutatorName": "ConditionalExpression",
    "replacement": "true",
  },
  Object {
    "location": "20:14-20:19",
    "mutatorName": "EqualityOperator",
    "replacement": "n <= 0",
  },
  Object {
    "location": "20:14-20:19",
    "mutatorName": "EqualityOperator",
    "replacement": "n >= 0",
  },
  Object {
    "location": "33:21-33:33",
    "mutatorName": "ArrowFunction",
    "replacement": "() => undefined",
  },
  Object {
    "location": "33:27-33:33",
    "mutatorName": "UpdateOperator",
    "replacement": "hits--",
  },
  Object {
    "location": "34:21-34:39",
    "mutatorName": "ArrowFunction",
    "replacement": "() => undefined",
  },
  Object {
    "location": "34:28-34:38",
    "mutatorName": "AssignmentOperator",
    "replacement": "hits -= 10",
  },
  Object {
    "location": "34:50-34:59",
    "mutatorName": "ArithmeticOperator",
    "replacement": "hits - 10",
  },
]
`;
